### ISSUE TITLE
[Bug]: test_spec_decode_logprobs never runs on NVIDIA CI, and is ill-defined at top-k ties when it does

### ISSUE THREAD

**Issue by jyan-R** [vllm-project/vllm#53620]
### Your current environment

<details>
<summary>Hardware and build</summary>

```text
GPU:    A100 80GB PCIe (SM80) and H100 PCIe (SM90)
Driver: 580.159.04
Torch:  2.13.0+cu132
vLLM:   upstream a3561ef8e, unpatched, VLLM_USE_PRECOMPILED=1
```

</details>

CI builds cross-checked: 83984 (my own PR #52311), and 85228 / 85276 / 85287 / 85298 / 85300.

Every experiment below ran against unpatched upstream `a3561ef8e`. I changed nothing in the test's requests, tolerances, assertions, or execution mode.

### 🐛 Describe the bug

`tests/v1/sample/test_logprobs.py::test_spec_decode_logprobs` has two independent problems:

1. **It never executes on NVIDIA CI.** `large_gpu_mark(min_gb=32)` gates all three model setups, but the step that hosts them is scheduled on an 18 GiB MIG slice. The measured peak footprint of the test is about 4 GiB.
2. **When it does execute, the `decoded_token` assertion has no defined answer.** Logprobs are paired by flat index rather than by token id, and exact ties at the top-k cut make the pairing depend on `torch.topk`'s unspecified tie-breaking.

The two are separable — fixing either one does not affect the other — so I have written them up separately.

---

# Part 1 — Why the test never runs

## 1.1 `min_gb=32` does not match what the test needs

All three model setups carry `large_gpu_mark(min_gb=32)`, while the NVIDIA step that hosts them, `(H200) V1 Sample + Logits`, runs on the `h200_18gb` queue (`.buildkite/test_areas/misc.yaml`). Since 18 < 32, all 12 parametrized nodes are skipped.

Two separate things stack up here:

- The one [#50463](https://github.com/vllm-project/vllm/pull/50463) is fixing: on a MIG partition NVML returns `Insufficient Permissions`, `large_gpu_mark` swallows the exception as 0GB, and the skip reason is rendered as `(found 0GB)`.
- The subject of this issue: the threshold value 32 is unrelated to the test's actual footprint.

Fixing either one alone leaves the test skipped. With only #50463, the probe correctly reports 18GB and `18 < 32` still skips; with only a threshold change, the probe still returns 0 and `0 < anything` still skips.

Both are visible in the same job from my own PR [#52311](https://github.com/vllm-project/vllm/pull/52311) (build [#83984](https://buildkite.com/vllm/ci/builds/83984), 2026-08-14):

```
/vllm-workspace/tests/v1/sample/test_logprobs.py:1080: UserWarning: An error occurred when finding the available memory: Insufficient Permissions
v1/sample/test_logprobs.py::test_spec_decode_logprobs[ngram-raw_logprobs] SKIPPED (Need at least 32GB GPU memory to run the test (found 0GB).)
================= 299 passed, 23 skipped, 25 warnings in 1038.74s =================
```

299 tests passed in that same job, so the GPU was usable.

## 1.2 Measured footprint: about 4 GiB

I ran these nodes on an A100 80GB and an H100 PCIe. The numbers below are taken from the `gpu_worker.py` memory breakdown vLLM prints at startup, not estimated:

| setup / engine | weights + non-torch | peak activation | CUDA graph | non-KV total |
|---|---|---|---|---|
| ngram / ref (A100) | 2.46 GiB | 0.48 GiB | 0.08 GiB | 3.02 GiB |
| ngram / spec (A100) | 2.94 GiB | 0.52 GiB | 0.04 GiB | 3.50 GiB |
| eagle3 / ref (A100) | 2.46 GiB | 0.01 GiB | 0.08 GiB | 2.55 GiB |
| eagle3 / spec (A100) | 2.75 GiB | 0.53 GiB | 0.12 GiB | 3.40 GiB |
| eagle3 / ref (H100) | 2.60 GiB | 0.01 GiB | 0.11 GiB | 2.72 GiB |
| eagle3 / spec (H100) | 3.34 GiB | 0.54 GiB | 0.18 GiB | **4.06 GiB** |

The raw log line:

```
Free memory on device (78.83/79.25 GiB) on startup. Desired GPU memory utilization is (0.4, 31.7 GiB).
Actual usage is 2.46 GiB for consumed memory (weights + non-torch), 0.48 GiB for peak activation,
and 0.08 GiB for CUDAGraph memory. ... Current kv cache memory in use is 28.75 GiB.
```

Across 3 setups and 2 architectures, the non-KV peak is **4.06 GiB**.

The ~28 GiB of KV cache is not a requirement — it is vLLM filling the budget it was handed. The test needs 2 requests x 256 tokens = 512 tokens of KV. The same log reports `GPU KV cache size: 942,224 tokens` for 28.75 GiB (about 30.5 KiB/token), which puts 512 tokens at roughly 16 MiB. The `Maximum concurrency for 256 tokens per request: 3680.56x` line in the same log agrees.

## 1.3 The test fits on the 18GB slice

`gpu_memory_utilization=0.4` is a fraction of total device memory:

| device | 0.4 budget | measured peak | headroom |
|---|---|---|---|
| H200 141GB (full card) | 56.4 GiB | ~4.1 GiB | 13x |
| A100 80GB (measured on) | 31.7 GiB | ~4.1 GiB | 7.7x |
| `h200_18gb` slice | 7.2 GiB | ~4.1 GiB | **1.8x** |

Even on the smallest slice there is close to 2x headroom, so `min_gb=32` is the only thing blocking the test, at roughly 8x the measured requirement.

---

# Part 2 — Why the assertion is ill-defined once it runs

I ran the 12 nodes on both machines, and separately captured the full pre-topk logits tensor for a bit-level comparison.

| config (`ngram` x `raw_logprobs` x `top_logprobs=3`) | A100 80GB PCIe | H100 PCIe |
|---|---|---|
| upstream default | FAIL | PASS |
| `max_num_seqs=1` | FAIL | PASS |
| `attention_backend=TRITON_ATTN` | FAIL | PASS |
| `max_num_seqs=1` + `TRITON_ATTN` | PASS | PASS |
| `compilation_config={"cudagraph_mode": "NONE"}` | FAIL | PASS |
| `enforce_eager=True` | PASS | PASS |
| upstream default, `top_logprobs=5` | FAIL | **FAIL** |

| config (A100, full pre-topk BF16 tensor, bit-level) | test | vocab entries differing | max \|Δ\| | mean \|Δ\| | max BF16 code steps within top-64 |
|---|---|---|---|---|---|
| upstream default | FAIL | 64.69% | 0.125 | 1.27e-02 | 2 |
| `cudagraph_mode=NONE` | FAIL | 66.08% | 0.125 | 1.36e-02 | 2 |
| `max_num_seqs=1` | FAIL | 77.23% | 0.125 | 1.54e-02 | 1 |
| `enforce_eager=True` | PASS | 68.44% | **0.1875** | **1.68e-02** | 1 |

From those two sets of data, the test appears to have the following problems, roughly in order of importance.

### (a) Entries are paired by flat index, not by token id

```python
ref_logprobs.extend(logprobs.values())    # flatten each position's dict in rank order
for ref_lp, spec_lp in zip(ref_logprobs, spec_logprobs):   # pair by index
```

The flat index is `(request, position, rank slot)`; the token id never participates in the pairing. As soon as the two engines' top-k membership differs, this compares ref's rank-3 against spec's rank-3, which are different tokens. All 8 of my A100 failures are exactly this case, with no exceptions.

### (b) The `decoded_token` assertion has no defined value at a tie

In every one of those 8 failures, the two tokens being compared have bit-identical raw BF16 logits in the spec engine.

These boundary ties are not rare: in every configuration, 1–4 of the 20 generated positions have an exact tie straddling the top-k cut, in both engines, with the largest tie group containing 3 tokens.

When there is a tie, which token lands at rank k depends on `torch.topk`'s unspecified tie-breaking behaviour, so **there is no correct answer to assert against**. This is independent of the score tolerance: the tolerance decides which assertion fires first, it cannot make the asserted quantity well-defined.

### (c) Pass/fail is not correlated with numerical quality

From the second table, `enforce_eager=True` has the largest max |Δ| and mean |Δ| of the four configurations and passes, while the upstream default has the smallest and fails. Turning off `torch.compile` does not reduce the divergence between the two paths; it moves the top-k cut off a tie.

The numbers themselves look healthy: at most 1 BF16 code step within the top-8, 1–2 within the top-64, 3 within the top-1024, and a whole-vocabulary absolute difference of ≤ 0.1875 against a logit range of roughly −6 to +23. That is close to the BF16 representation limit.

I have not located the source of this difference (which kernel, which accumulation order), and I did not run a layer-by-layer probe.

### (d) `GPU_DETERMINISM_KWARGS` is empty on CUDA

```python
if current_platform.is_rocm():
    GPU_DETERMINISM_KWARGS = dict(max_num_seqs=1, attention_backend="TRITON_ATTN")
elif current_platform.is_xpu():
    GPU_DETERMINISM_KWARGS = dict(max_num_seqs=1, attention_backend="FLASH_ATTN")
else:
    GPU_DETERMINISM_KWARGS = {}
```

The determinism controls the test relies on exist only on the ROCm and XPU branches. Per the first table, transplanting them to CUDA does not fix it either: `max_num_seqs=1` alone raises the mismatch count from 1 to 2, `TRITON_ATTN` alone still fails, and only the combination passes. These knobs move the cut line rather than repair a computation.

### (e) The failure message prints only the ref-side token

```python
f"diff={abs(ref_logprob.logprob - spec_logprob.logprob)} "
f"(token={ref_logprob.decoded_token!r})"
```

Combined with (a): when the two sides hold different tokens, that `diff` is the gap between two different tokens' logprobs rather than the error on one token — and in the log the two cases are indistinguishable. This makes historical "logprob mismatch" reports against this test hard to interpret after the fact.

### (f) The docstring claims eager mode, which the implementation never sets

> Both use identical chunked prefill settings and eager mode to control for
> infrastructure differences.

`llm_kwargs` never sets `enforce_eager`, so both LLMs run with compile + CUDA graphs. On its own this is a documentation mismatch, but combined with (c) it is worth more than a typo: eager mode does change the verdict, just not for the reason the docstring implies.

### What is not broken

In every configuration, including all the failing ones, the sampled output tokens from the reference and speculative engines are identical. What speculative decoding actually promises holds; the property that goes red is one it never promised.

---

# Proposed fixes

**Part 1**

1. Lower `min_gb` to a value commensurate with the measured footprint. From the numbers above, `min_gb=8` gives 2x headroom and `min_gb=12` gives 3x, and either lets the test run on `h200_18gb`. I have no strong view on the exact value, only that 32 is not derived from this test.
2. If 32 was chosen for a reason other than this test's footprint (multi-GPU, larger models), recording that in the code would help — otherwise it is not possible to tell from the outside whether the value is safe to change.
3. This should land alongside #50463: the threshold change makes the test runnable on a qualifying runner, and #50463 stops the probe from misreporting. Neither is sufficient alone.

**Part 2**

4. Pair by token id per position instead of zipping by flat index: for tokens present in both engines' top-k, assert the logprob is within the existing tolerance and the rank matches; assert exact equality on the sampled token sequence (speculative decoding's actual contract, never violated in any of my configurations); and allow the two top-k sets to differ only within a boundary tie group.
5. Whichever approach is taken, print both the ref and spec tokens in the failure message. It is a one-line change, and it decides whether future failure reports are interpretable.

# Expected CI impact

Lowering the threshold will not turn CI red. The current NVIDIA runners are H200, the same Hopper generation as my H100, and Hopper passes at the `top_logprobs=3` the test actually uses (first table in Part 2). The failure I see on A100 needs `top_logprobs=5` to surface on Hopper.

Part 1 and Part 2 are therefore independent in both directions: fixing the threshold will not expose the assertion problem, and fixing the assertion will not make the test run.

# Related issues

[#46657](https://github.com/vllm-project/vllm/issues/46657) discusses the same test. It reports a score-tolerance failure on ROCm, which is a different signature from the `decoded_token` failure here, and I am not claiming a shared root cause.

Per the duplicate-work check in `AGENTS.md`, I searched open issues and PRs for `large_gpu_mark`, `min_gb`, `found 0GB`, and `test_spec_decode_logprobs`. The related open items are #50463 (fixes the probe, leaves the threshold), #41776 and #35470 (MIG UUID resolution), and #46657 (the ROCm report). None of them addresses the threshold value or the assertion design.

AI assistance was used for the analysis and for writing the measurement scripts. All experiments were run by me on real GPUs, and every number above was checked against the raw logs.

**Comment by sudhirpol522** [comment]
I’m working on a test-only fix for this issue. My plan is to:

Lower the GPU-memory gate to 16 GiB, which provides a 6.4 GiB allocation at gpu_memory_utilization=0.4.
Compare logprobs per position by token ID instead of dictionary order.
Allow top-k membership/rank differences only around exact cutoff ties while requiring identical sampled token sequences.

I’ll keep #50463 as a dependency rather than duplicating its MIG memory-probe fix. I found no open PR covering these changes.
